### PR TITLE
[TRANSFORMATIONS] Fix broacasting check for MultiplyConvolutionFusion

### PR DESCRIPTION
**Pull request by CuriousPanCake**
[TRANSFORMATIONS] Fix broadcasting check for MultiplyConvolutionFusion

The previous implementation of broadcasting checks in the MultiplyConvolutionFusion transformation
was not covering all the possible cases.
Add a check to eliminate cases when input is broadcasted by constant in multiply.

GitHub issue: https://github.com/openvinotoolkit/openvino/issues/25649

- Tickets:
	* CVS-148621

Signed-off-by: Andrii Staikov <andrii.staikov@intel.com>

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp [hunk 1]
```diff
@@ -41,11 +71,12 @@ ov::pass::MultiplyConvolutionFusion::MultiplyConvolutionFusion() {
 
         const auto& weights_shape = weights.get_shape();
         const auto& mul_const_shape = mul_const.get_shape();
-        // Check if mul_const if broadcastable to weights.
+
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp:74]: @mateusztabaka is also suggesting to add a following check here:
```
        if (input.get_partial_shape().size() < weights_shape.size())
            return false;
```
but I don't really understand what's the point of it here.

File: src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp [hunk 2]
```diff
@@ -41,11 +71,12 @@ ov::pass::MultiplyConvolutionFusion::MultiplyConvolutionFusion() {
...
+        // Check if mul_const is broadcastable to weights.
         // Also if mul_const's rank matches weights rank and mul_const.shape[0] != 1
         // then we can't fuse the multiply, since first dimension in mul_const corresponds to
         // batch size, while first dimension in weights corresponds to output channel count
-        if (op::util::check_for_broadcast(weights_shape, mul_const_shape) ||
+        if (!are_shapes_compatible(weights_shape, mul_const_shape) ||
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp:79]: `op::util::check_for_broadcast` is also used in other places
please replace all usages with a new correct version.

File: src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp (hunk 1 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp:75]: Added

File: src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp (hunk 2 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp]: fixed.

File: src/common/transformations/src/transformations/utils/utils.cpp [hunk 3]
```diff
@@ -155,6 +155,35 @@ bool check_for_broadcast(const ov::PartialShape& ref_shape, const ov::PartialSha
...
 }
 
+bool broadcasted_only_channel(const ov::PartialShape& weights_shape, const ov::PartialShape& const_shape) {
+    if (const_shape.size() > weights_shape.size()) {
+        return false;
+    }
+
+    auto weights_shape_it = weights_shape.rbegin();
+    auto const_shape_it = const_shape.rbegin();
+    size_t C_dim = const_shape.size() - 1 - 2;
```
Comment (itikhono) [inline src/common/transformations/src/transformations/utils/utils.cpp:165]: what if input is 3d or 5d?
it looks like this calculation will break `C_dim = const_shape.size() - 1 - 2;`

File: src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp [hunk 4]
```diff
@@ -106,7 +111,7 @@ ov::pass::MultiplyGroupConvolutionFusion::MultiplyGroupConvolutionFusion() {
             // Reshape mul_const from shape (1, C, H, W) to (G, 1, C / G, H, W) to match GroupConvolution weights format
             Shape new_shape{G, 1, C};
             std::copy(mul_const_shape.begin() + 2, mul_const_shape.end(), std::back_inserter(new_shape));
-            if (op::util::check_for_broadcast(weights_shape, new_shape)) {
+            if (!ov::op::util::broadcasted_only_channel(weights_shape, new_shape)) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/mul_conv_fusion.cpp:114]: it looks like, for Mul -> Conv case we have to check not only the channel dim
so please revert the changes
